### PR TITLE
RFC for Issue #306

### PR DESCRIPTION
**Pull request by michael-mclawhorn**
https://github.com/yahoo/fili/issues/306

### CODE REVIEW

**Comment by archolewa** [comment]
If we're going to do this, then we may want to add an additional `ApiRequest`, `FilterableApiRequest` that is an `ApiRequest` with filters, that the `Dimensions`, `Data` and `Table` requests can all extend from. As soon as you add filters to a request, there opens up the doors to being able to add implicit filters. Having to either duplicate the implicit filter for each concrete request, or making heavy use of `instanceof` and casting are mildly annoying with two filterable requests. Adding a third makes it even worse.

**Comment by michael-mclawhorn** [comment]
@archolewa That makes sense.  

One motivating desire to leverage the query planning capabilities of a PartitionedTable by applying ApiFilters to the query should go under the same query rewrite restrictions.

We could roll this out with no api support for filters to simplify security concerns.  But internally we definitely want to be able to attach filters.  It's possible, in conjunction with your suggestion, that the right thing to do would be to have the ApiRequest do a lookup for any pertinent security filters from a registry, and rewrite itself with any filters that match it.

**Comment by michael-mclawhorn** [comment]
e.g. If it's an APIRequest that accepts FilterValidation, it should get all security rules that target FilterValidation and map through each of them.

We really need to figure out the default security model sooner than later.

**Comment by michael-mclawhorn** [comment]
@archolewa I've got a change about to be in a PR where I create an additional ChainingApiRequestMapper to support richer security support.
This isn't directly pertinent to your recommendation about commonality on 'has filter' being in the type structure of the ApiRequest, but I thought you might want to know, as you have the largest current use of transforming ApiRequests.

File: rfc/tableAvailabilityMarkdown.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+```
+
+The table metadata endpoint should now accept an optional list of path separated grouping dimensions, an optional list
+of metrics, and an optional filter clause.
+
+Code currently in `DataApiRequest` will be shared with `TablesApiRequest` to support parsing these elements.
+
+## Response Format Changes:
+
+Add `availability` as a list of serialized intervals to the default table format.
```
Comment (cdeszaq) [inline rfc/tableAvailabilityMarkdown.md]: Is this the label for the new field? It doesn't match the example...

**Comment by cdeszaq** [comment]
This looks pretty good to me. Here are the general thoughts I have:

1. We should leverage as much of the existing code in DataApiRequest as we can. And rather than inheritance, we should look more at composing I think. We're already pretty inherit-centric in many places and it's caused more pain than I think we anticipated as we've tried to change things. Perhaps moving the common logic into helper utilities would get us enough separation and re-use?

2. Along with code reuse, I think leaning on interfaces for "this thing can have filters added to it via a  `with*` method is probably a better route than inheriting from a common parent.

3. We should definitely figure out a general security model / framework, starting with the needs. I'll try to start a discussion in the dev group and likely an issue as well. But let's not hold this effort here up on not having a stronger sense of security needs.

**Review by cdeszaq — APPROVED** [review]

**Comment by QubitPi** [comment]
I can implement this RFC.

File: rfc/tableAvailabilityMarkdown.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
+# Table resource supporting filtered availability
+
+This RFC implements [Github Issue 306](https://github.com/yahoo/fili/issues/306)
+
+## Goals
+
+1. Show the queryable range of time as part of the Logical Table endpoint view.
+2. Allow the queryable range to reflect constraints as would be found in a data request.
+  * Required metrics and dimensions
+  * Applied filters
```
Comment (aklish) [inline rfc/tableAvailabilityMarkdown.md]: Do we really need filters?  It seems like the only information they convey are a set of dimensions that are included in the filter request.  The API to return availability seems odd having to specify data filters.  Does the operator matter?  What about the filter values?  Would I get a different result back if I use startsWith vs contains?

Comment (michael-mclawhorn) [inline rfc/tableAvailabilityMarkdown.md]: The filters can impact availability if dimension /values/ impact the query planning.  (right now that's only a concern on physical tables with dimension value partitions)

File: rfc/tableAvailabilityMarkdown.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+1. Build a `QueryPlanningConstraint` from the TableApiRequest
+    From: `DruidQueryBuilder`
+    ```java 
+        TableGroup group = logicalTable.getTableGroup();
+    
+        // Resolve the table from the the group, the combined dimensions in request, and template time grain
+        QueryPlanningConstraint constraint = new QueryPlanningConstraint(request, template);
+    
+    ```
+    * In the case of unconstrained queries, ensure that an 'unconstrained' constraint is built.
```
Comment (QubitPi) [inline rfc/tableAvailabilityMarkdown.md]: How do you know if a query is unconstrained?

**Review by ghost — COMMENTED** [review]
Looks like 1. 2. 3. have already been implemented. Are we implementing a set of new classes that utilizes those?

File: rfc/tableAvailabilityMarkdown.md (hunk 3 above)
Comment (michael-mclawhorn) [inline rfc/tableAvailabilityMarkdown.md]: By unconstrainted I mean that it's constrained only by fields specified in the ApiRequest.  So if the ApiRequest doesn't list dimensions, no dimensional constraints are applied, and so forth.

**Comment by michael-mclawhorn** [comment]
I'm not sure where you mean 1 is implemented.  Can you clarify where the code in question is?

**Comment by QubitPi** [comment]
By 1, 2, 3, I mean 

**1**. Build a `QueryPlanningConstraint` from the TableApiRequest
    From: `DruidQueryBuilder`
    ```java 
        TableGroup group = logicalTable.getTableGroup();
    
        // Resolve the table from the the group, the combined dimensions in request, and template time grain
        QueryPlanningConstraint constraint = new QueryPlanningConstraint(request, template);
    
    ```
    * In the case of unconstrained queries, ensure that an 'unconstrained' constraint is built.
    * `QueryPlanningConstraint` will need more constructor richness to handle making metric constraints optional
    
**2**.  Get the tables from the `LogicalTable`'s `tableGroup`.
**3**.  Use matchers to filter the tables from the table group given the `QueryPlanningConstraint`
From: `DefaultPhysicalTableResolver`
    ```java
        public List<PhysicalTableMatcher> getMatchers(QueryPlanningConstraint requestConstraint) {
            return Arrays.asList(
                    new SchemaPhysicalTableMatcher(requestConstraint),
                    new AggregatableDimensionsMatcher(requestConstraint),
                    new TimeAlignmentPhysicalTableMatcher(requestConstraint)
            );
        }
    ```

    From: `BasePhysicalTableResolver`
    ```java
        public Set<PhysicalTable> filter(
                Collection<PhysicalTable> candidateTables,
                QueryPlanningConstraint requestConstraint
        ) throws NoMatchFoundException {
            return filter(candidateTables, getMatchers(requestConstraint));
        }
    ```

**Comment by michael-mclawhorn** [comment]
This was implemented

**Pull request closed without merging**